src/test/java/com/fasterxml/jackson/databind/introspect/MethodGenericTypeResolverTest.java

```diff
@@ -0,0 +1,196 @@
+package com.fasterxml.jackson.databind.introspect;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.core.type.TypeReference;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.type.TypeBindings;
+import com.fasterxml.jackson.databind.type.TypeFactory;
+import org.junit.Test;
+
+import java.lang.reflect.Method;
+import java.lang.reflect.Modifier;
+import java.lang.reflect.Type;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class MethodGenericTypeResolverTest {
+
+    private static final TypeResolutionContext EMPTY_CONTEXT =
+            new TypeResolutionContext.Empty(TypeFactory.defaultInstance());
+
+    public static <T> Optional<T> simple(T input) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static Optional<?> noGenerics(String input) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static <T> Map<T, T> mapWithSameKeysAndValues(List<T> input) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static <T> Map<?, ?> disconnected(List<T> input) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static <A, B> Map<A, B> multipleTypeVariables(Map<A, B> input) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static <A, B> Map<? extends A, ? extends B> multipleTypeVariablesWithUpperBound(Map<A, B> input) {
+        throw new UnsupportedOperationException();
+    }
+
+    public static class StubA {
+        private final String value;
+
+        @JsonCreator
```

Comment:
I would suggest adding explicit `(mode = JsonCreator.Mode.DELEGATING)` here, as this is potentially ambiguous case (esp. as parameter name and field are both `value`) and could be detected as properties-based.